relentless/optimize/method.py

```diff
@@ -0,0 +1,49 @@
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        pass
+
+class SteepestDescent(Optimizer):
+    def __init__(self, alpha, max_iter, abs_tol, mode):
+        self.alpha = alpha
+        self.max_iter = max_iter
+        self.abs_tol = abs_tol
+        self.mode = mode
+
+    def optimize(self, objective):
+        dvars = objective.design_variables()
+        iter_num = 1
+        update = {}
+
+        while True:
+            res = objective.compute()
+            for x in dvars:
+                update[x] = self.alpha*res.gradient(x)
+
+            if iter_num > self.max_iter:
+                raise RuntimeError('Could not find solution within ' + str(self.max_iter) + ' iterations.')
+            elif self._has_converged(update):
```

Comment:
I think I would still always apply the update, even if the method has converged, because it shouldn't hurt to advance that remaining small bit.